### ISSUE TITLE
ACL denying subscribes on wildcards to prevent exponential overload

### ISSUE THREAD

**Issue by wiebeytec** [eclipse/mosquitto#412]
When I subscribe on `#` or something like `N/#`, I don't get that request in the auth plugin function `mosquitto_auth_acl_check`. This is postponed till when publishes come in, and they are then verified against _every_ wildcard subscriber. This causes a huge and unnecessary exponential load on the server: take a client with 1000 topics: every new client that subscribes needs to have the `number_of_clients * 1000` verified, and also introduces 1000 of its own.

Our server is having a hard time right now because one person subscribed to `N/#`.

I would like the `mosquitto_auth_acl_check` to be called on a wildcard subscribe to be able to reject these from the onset. (In my case, wildcards are allowed, but only after a few segments (like `N/123456/#`), so that I can use the 123456 to test ACL.)

**Comment by karlp** [comment]
I feel your pain, but by spec N/# should match the portions of N/blah/whatever that they are allowed to subscribe to.

I don't see how you can reject them ahead of time, at least not in any spec compliant way.

**Comment by wiebeytec** [comment]
Hmm. So can I DOS any MQTT server by subscribing to `#` a bunch of times? Perhaps sometimes the spec doesn't suffice. Another idea would be to only allow wildcards after the nth slash, but I guess that's non-standard as well...

Anyway, Mosquitto does caching. What is the algorithmic complexity of its key lookup? My new auth plugin will be a hash map for ACL checking, so will be virtually constant time, but the majority of the requests will be handled by Mosquitto's cache, and even `log n` may be too slow.

**Comment by karlp** [comment]
wiebeytec <notifications@github.com> wrote:
> Hmm. So can I DOS any MQTT server by subscribing to `#` a bunch
> of times? 

no, only those that implement their acl checks in certain ways :)

I'm not saying you can't implement something that performs
better, but "rejecting wild card subscriptions" ahead of time
can't be the way, because the wildcard subscription can overlap
with _allowed_ subscriptions.

Perhaps sometimes the spec doesn't suffice. Another
> idea would be to only allow wildcards after the nth slash, but
> I guess that's non-standard as well...
> 
> Anyway, Mosquitto does caching. What is the algorithmic
> complexity of its key lookup? My new auth plugin will be a hash
> map for ACL checking, so will be virtually constant time, but
> the majority of the requests will be handled by Mosquitto's
> cache, and even `log n` may be too slow.
>

**Comment by wiebeytec** [comment]
I will bite the bullet and close this, as it would be non-standard. Luckily, I have some ideas on how to make the ACL check very efficient, so I can mitigate the situation.
